### PR TITLE
CB-21512 basic framework for secret rotation

### PR DESCRIPTION
**Pull request by horadla23**
- generating context for secret rotation
- executing rotation based on type of secret
- WIP: error handling, rollback strategy, DB secret rotations
- added rotations: CM admin password, CM mgmt password
- not wired into any API or automatism yet

See detailed description in the commit message.

### CODE REVIEW

File: common/src/main/java/com/sequenceiq/cloudbreak/rotation/secret/PostRotationAction.java [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package com.sequenceiq.cloudbreak.rotation.secret;
+
+public enum PostRotationAction {
+    REPAIR,
+    NOTHING
```
Comment (lnardai) [inline common/src/main/java/com/sequenceiq/cloudbreak/rotation/secret/PostRotationAction.java:5]: maybe a "NO_ACTION" is more descriptive.

File: common/src/main/java/com/sequenceiq/cloudbreak/rotation/secret/RotationType.java [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package com.sequenceiq.cloudbreak.rotation.secret;
+
+public enum RotationType {
```
Comment (lnardai) [inline common/src/main/java/com/sequenceiq/cloudbreak/rotation/secret/RotationType.java:3]: I think this somewhat overlaps with the SecretType.
This enum is would rather represent the location where the secret is stored right?
How about SecretStorageType?
Them the CM_USER would become CM_SERVER.

Comment (horadla23) [reply common/src/main/java/com/sequenceiq/cloudbreak/rotation/secret/RotationType.java:3]: We can rename the class, I have been thinking about naming a lot, it is hard to differentiate easily the secret we rotate and the place we are rotating it. CM_SERVER would be though too general, probably several other secrets can be named like that too. We should clearly indicate what type of secret we are rotating.
For example we will rotate DB password and CM user password too on CM server.

Comment (horadla23) [reply common/src/main/java/com/sequenceiq/cloudbreak/rotation/secret/RotationType.java:3]: > This enum is would rather represent the location where the secret is stored right?
yes. maybe SecretLocationType

File: common/src/main/java/com/sequenceiq/cloudbreak/rotation/secret/RotationExecutor.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package com.sequenceiq.cloudbreak.rotation.secret;
+
+public interface RotationExecutor {
+
+    void rotate(RotationContext rotationContext);
```
Comment (lnardai) [inline common/src/main/java/com/sequenceiq/cloudbreak/rotation/secret/RotationExecutor.java:5]: would it make sense to add a rollback method here also?

File: secret-engine/src/main/java/com/sequenceiq/cloudbreak/service/secret/service/SecretService.java [hunk 4]
```diff
@@ -66,22 +65,19 @@ public void init() {
...
         String secret = vaultRetryService.tryWritingVault(() -> persistentEngine.put(key, value));
-        duration = System.currentTimeMillis() - start;
+        long duration = System.currentTimeMillis() - start;
         metricService.gauge(MetricType.VAULT_WRITE, duration);
         LOGGER.trace("Secret write took {} ms", duration);
         metricService.incrementMetricCounter(() -> "secret.write." + convertSecretToMetric(secret));
         return secret;
     }
 
+    public String update(String secret, String newValue) throws Exception {
```
Comment (lnardai) [inline secret-engine/src/main/java/com/sequenceiq/cloudbreak/service/secret/service/SecretService.java:76]: Would this automatically create a v2 version of the secret?

Comment (horadla23) [reply secret-engine/src/main/java/com/sequenceiq/cloudbreak/service/secret/service/SecretService.java:76]: yes

File: common/src/main/java/com/sequenceiq/cloudbreak/rotation/secret/RotationExecutor.java (hunk 3 above)
Comment (horadla23) [reply common/src/main/java/com/sequenceiq/cloudbreak/rotation/secret/RotationExecutor.java:5]: yes, i have not started to add rollback related part of the framework, can be done when we start to work on it

File: secret-engine/src/main/java/com/sequenceiq/cloudbreak/service/secret/service/SecretService.java (hunk 4 above)
Comment (horadla23) [reply secret-engine/src/main/java/com/sequenceiq/cloudbreak/service/secret/service/SecretService.java:76]: sure, good idea

**Comment by lnardai** [comment]
Would it make sense or be possible to write mock tests for secret updates? In theory we have all the bits mocked already.
If it's not a big chore I would add an example test with the framework.

**Comment by horadla23** [comment]
> Would it make sense or be possible to write mock tests for secret updates? In theory we have all the bits mocked already. If it's not a big chore I would add an example test with the framework.

there is no invocation of the framework yet

**Review by lnardai — APPROVED** [review]

**Review by sodre90 — APPROVED** [review]

**Comment by horadla23** [comment]
rerequested review only for make the PR visible on master merge list, there is nothing new to review

**Pull request merged**